### PR TITLE
varlink containers

### PR DESCRIPTION
**Pull request by baude**
first pass at adding in the container related endpoints/methods for the libpod
backend. Couple of important notes:

* endpoints that can use a console are not going to be done until we have "remote" console
* several of the container methods should probably be able to stream as opposed to a one-off return

Signed-off-by: baude <bbaude@redhat.com>

### CODE REVIEW

**Comment by baude** [comment]
[(5) podman varlink containers API](https://trello.com/c/FZoEWubs/567-5-podman-varlink-containers-api)

File: libpod/container_api.go [hunk 1]
```diff
@@ -705,3 +705,19 @@ func (c *Container) Sync() error {
 
 	return nil
 }
+
+// Restart restarts a running container and takes a given timeout in uint
+func (c *Container) Restart(timeout uint) error {
+	ctx := context.TODO()
```
Comment (mheon) [inline libpod/container_api.go]: Pass this in as a parameter

File: libpod/container_api.go [hunk 2]
```diff
@@ -705,3 +705,19 @@ func (c *Container) Sync() error {
...
+	state, err := c.State()
```
Comment (mheon) [inline libpod/container_api.go]: Don't use this - lock the container and do a syncContainer()

File: libpod/container_api.go [hunk 3]
```diff
@@ -705,3 +705,19 @@ func (c *Container) Sync() error {
...
+	if err != nil {
+		return err
+	}
+	if state == ContainerStateRunning {
+		if err := c.StopWithTimeout(timeout); err != nil {
```
Comment (mheon) [inline libpod/container_api.go]: Use internal non-locking functions

File: libpod/container_api.go [hunk 4]
```diff
@@ -705,3 +705,19 @@ func (c *Container) Sync() error {
...
+			return err
+		}
+	}
+
+	return c.Start(ctx)
```
Comment (mheon) [inline libpod/container_api.go]: Use internal non-locking functions

File: libpod/container_api.go [hunk 5]
```diff
@@ -705,3 +705,19 @@ func (c *Container) Sync() error {
 
 	return nil
 }
+
+// Restart restarts a running container and takes a given timeout in uint
+func (c *Container) Restart(timeout uint) error {
```
Comment (mheon) [inline libpod/container_api.go]: Call this RestartWithTimeout(), our standard Stop() doesn't take a timeout

File: cmd/podman/varlink/io.projectatomic.podman.varlink [hunk 6]
```diff
@@ -53,32 +53,71 @@ type ImageSearch (
...
+    block_output: int,
+    block_input: int,
+    pids: int
+)
+
 # System
 method Ping() -> (ping: StringResponse)
 method GetVersion() -> (version: Version)
 
 # Containers
```
Comment (mheon) [inline cmd/podman/varlink/io.projectatomic.podman.varlink]: Do we have an API target to get a single container by ID?

File: libpod/container_api.go (hunk 1 above)
Comment (baude) [reply libpod/container_api.go]: fixed

File: libpod/container_api.go (hunk 5 above)
Comment (baude) [reply libpod/container_api.go]: fixed

File: libpod/container_api.go (hunk 2 above)
Comment (baude) [reply libpod/container_api.go]: fixed

**Comment by mheon** [comment]
Would be nice if we had two versions of ListContainers - one that only did a single container (pass it a name or ID), one that did all containers

File: libpod/container_api.go (hunk 3 above)
Comment (baude) [reply libpod/container_api.go]: fixed

File: libpod/container_api.go (hunk 4 above)
Comment (baude) [reply libpod/container_api.go]: fixed

**Comment by baude** [comment]
agree, will do.

File: libpod/container_api.go [hunk 7]
```diff
@@ -705,3 +705,23 @@ func (c *Container) Sync() error {
 
 	return nil
 }
+
+// Restart restarts a running container and takes a given timeout in uint
```
Comment (mheon) [inline libpod/container_api.go]: Lint probably going to complain this isn't `// RestartWithTimeout restarts...`

File: cmd/podman/varlink/io.projectatomic.podman.varlink [hunk 8]
```diff
@@ -53,32 +53,72 @@ type ImageSearch (
...
+# ListContainer is the returned struct for an individual container
+type ListContainerData (
+    id: string,
+    image: string,
+    imageid: string,
+    command: []string,
+    createdat: string,
+    runningfor: string,
+    status: string,
+    ports: [][string]string,
```
Comment (mheon) [inline cmd/podman/varlink/io.projectatomic.podman.varlink]: Can we either do arrays of structs, or somehow expose the keys we expect to be in every map? So people know how to parse the ports and mounts?

File: libpod/container_api.go (hunk 7 above)
Comment (baude) [reply libpod/container_api.go]: nice catch!

File: cmd/podman/varlink/io.projectatomic.podman.varlink (hunk 6 above)
Comment (baude) [reply cmd/podman/varlink/io.projectatomic.podman.varlink]: per our IRC conversation, I added one

File: cmd/podman/varlink/io.projectatomic.podman.varlink (hunk 8 above)
Comment (baude) [reply cmd/podman/varlink/io.projectatomic.podman.varlink]: i took a todo from our irc argument about this.

**Comment by baude** [comment]
udpated with structs and no containers functions return empty responses.

File: cmd/podman/varlink/io.projectatomic.podman.varlink [hunk 9]
```diff
@@ -53,32 +53,100 @@ type ImageSearch (
...
+method GetContainer(name: string) -> (container: ListContainerData)
 method CreateContainer() -> (notimplemented: NotImplemented)
-method InspectContainer() -> (notimplemented: NotImplemented)
-method ListContainerProcesses() -> (notimplemented: NotImplemented)
-method GetContainerLogs() -> (notimplemented: NotImplemented)
-method ListContainerChanges() -> (notimplemented: NotImplemented)
-method ExportContainer() -> (notimplemented: NotImplemented)
-method GetContainerStats() -> (notimplemented: NotImplemented)
+method InspectContainer(name: string) -> (container: string)
+# TODO: Should this be made into a streaming response as opposed to a one off?
```
Comment (mheon) [inline cmd/podman/varlink/io.projectatomic.podman.varlink]: IMO safest to just have the client call this repeatedly as opposed to supporting proper streaming through the API.

File: libpod/container_api.go [hunk 10]
```diff
@@ -705,3 +705,23 @@ func (c *Container) Sync() error {
...
+			return err
+		}
+	}
+
+	if c.state.State == ContainerStateRunning {
+		if err := c.stop(timeout); err != nil {
+			return err
+		}
+	}
+
```
Comment (mheon) [inline libpod/container_api.go]: If c.state.State is ContainerStateConfigured we need to c.init() here

File: libpod/container_api.go [hunk 11]
```diff
@@ -705,3 +705,23 @@ func (c *Container) Sync() error {
...
+func (c *Container) RestartWithTimeout(ctx context.Context, timeout uint) error {
+	if !c.batched {
+		c.lock.Lock()
+		defer c.lock.Unlock()
+
+		if err := c.syncContainer(); err != nil {
+			return err
+		}
+	}
+
```
Comment (mheon) [inline libpod/container_api.go]: If c.state.State is ContainerStateUnknown or ContainerStatePaused we should return an error

**Comment by mheon** [comment]
Podman restart tests are failing. Dropped some comments that might help.

File: libpod/container_api.go (hunk 10 above)
Comment (mheon) [reply libpod/container_api.go]: Also, after this, if c.state.State is ContainerStateStopped, we need to https://github.com/projectatomic/libpod/blob/master/libpod/container_api.go#L103-L106

File: cmd/podman/varlink/io.projectatomic.podman.varlink [hunk 12]
```diff
@@ -53,32 +53,72 @@ type ImageSearch (
...
 method RenameContainer() -> (notimplemented: NotImplemented)
-method PauseContainer() -> (notimplemented: NotImplemented)
-method UnpauseContainer() -> (notimplemented: NotImplemented)
+method PauseContainer(name: string) -> ()
+method UnpauseContainer(name: string) -> ()
 method AttachToContainer() -> (notimplemented: NotImplemented)
-method WaitContainer() -> (notimplemented: NotImplemented)
-method RemoveContainer() -> (notimplemented: NotImplemented)
-method DeleteStoppedContainers() -> (notimplemented: NotImplemented)
+method WaitContainer(name: string) -> (exitcode: int)
```
[the remark this answers is not in the document]
Comment (baude) [reply cmd/podman/varlink/io.projectatomic.podman.varlink]: I dont think so .. but open to change.

File: cmd/podman/batchcontainer/container.go [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+		ConState:    conState,
+		ExitCode:    exitCode,
+		Pid:         pid,
+		RootFsSize:  rootFsSize,
+		RwSize:      rwSize,
+		StartedTime: startedTime,
+	}, nil
+}
+
+// GetNamespaces retunrs a populated namespace struct
```
[the remark this answers is not in the document]
Comment (baude) [reply cmd/podman/batchcontainer/container.go]: fixed

File: libpod/container_api.go (hunk 11 above)
Comment (baude) [reply libpod/container_api.go]: fixed

**Comment by mheon** [comment]
LGTM

**Comment by rhatdan** [comment]
I will wait for @jwhonce approval before merging, since he is the one having to use them.

File: cmd/podman/varlink/io.projectatomic.podman.varlink [hunk 14]
```diff
@@ -53,32 +53,100 @@ type ImageSearch (
...
+    names: string,
+    labels: [string]string,
+    mounts: []ContainerMount,
+    containerrunning: bool,
+    namespaces: ContainerNameSpace
+)
+
+# ContainerStats is the return struct for the stats of a container
+type ContainerStats (
+    container_id: string,
```
Comment (jwhonce) [inline cmd/podman/varlink/io.projectatomic.podman.varlink]: I would prefer 'id' vs. 'container_id' here.

File: cmd/podman/varlink/io.projectatomic.podman.varlink [hunk 15]
```diff
@@ -53,32 +53,100 @@ type ImageSearch (
...
+    net_output: int,
+    block_output: int,
+    block_input: int,
+    pids: int
+)
+
+# ContainerMount describes the struct for mounts in a container
+type ContainerMount (
+    destination: string,
+    type: string,
```
Comment (jwhonce) [inline cmd/podman/varlink/io.projectatomic.podman.varlink]: If type is an enum, can the valid list of values be included in the comments?  Or, is that future work with the auto-documentation?

File: cmd/podman/varlink/ioprojectatomicpodman.go [hunk 16]
```diff
@@ -37,172 +92,228 @@ type ImageSearch struct {
...
 }
 
-type Version struct {
-	Version    string `json:"version"`
-	Go_version string `json:"go_version"`
-	Git_commit string `json:"git_commit"`
-	Built      int64  `json:"built"`
-	Os_arch    string `json:"os_arch"`
+type ContainerStats struct {
+	Container_id string  `json:"container_id"`
```
Comment (jwhonce) [inline cmd/podman/varlink/ioprojectatomicpodman.go]: Preference is 'id' vs. 'container_id'

File: cmd/podman/varlink/ioprojectatomicpodman.go [hunk 17]
```diff
@@ -815,6 +961,14 @@ func (c__ *VarlinkCall) ReplyImageNotFound(imagename_ string) error {
 	return c__.ReplyError("io.projectatomic.podman.ImageNotFound", &out)
```
Comment (jwhonce) [inline cmd/podman/varlink/ioprojectatomicpodman.go]: I would prefer 'Imagename' to be one of 'id'/'name'/'image'

File: cmd/podman/varlink/ioprojectatomicpodman.go [hunk 18]
```diff
@@ -815,6 +961,14 @@ func (c__ *VarlinkCall) ReplyImageNotFound(imagename_ string) error {
...
 }
 
+func (c__ *VarlinkCall) ReplyContainerNotFound(containername_ string) error {
+	var out struct {
+		Containername string `json:"containername"`
```
Comment (jwhonce) [inline cmd/podman/varlink/ioprojectatomicpodman.go]: I would prefer 'containername' to be one of 'id'/'name'/'container'

File: cmd/podman/varlink/ioprojectatomicpodman.go [hunk 19]
```diff
@@ -1519,32 +1792,100 @@ type ImageSearch (
...
+    names: string,
+    labels: [string]string,
+    mounts: []ContainerMount,
+    containerrunning: bool,
+    namespaces: ContainerNameSpace
+)
+
+# ContainerStats is the return struct for the stats of a container
+type ContainerStats (
+    container_id: string,
```
Comment (jwhonce) [inline cmd/podman/varlink/ioprojectatomicpodman.go]: I would prefer 'id' over 'container_id'

**Comment by jwhonce** [comment]
LGTM :+1:

**Comment by baude** [comment]
@rh-atomic-bot r+

**Pull request closed without merging**